File: blender_extension/roblox_interface/roblox_server/format_types/buf.py
import struct

format = struct.Struct("<I")
format_size = format.size
unpack_from = format.unpack_from
pack = format.pack

def read(self, format_data, **masks):
	buf_len, = unpack_from(*self.__read_buffer(format_size))
	read_buffer, read_offset = self.__read_buffer(buf_len)
	return (read_buffer[read_offset:read_offset + buf_len],)

def write(self, args, args_count, format_data, **masks):
	buf_value = args[args_count]
	data_size = len(buf_value)
	self.__write_buffer(pack(data_size), format_size)
	self.__write_buffer(buf_value, data_size)
	return args_count + 1

File: roblox_plugin/src/Server/formatTypes/buf.ts
//!native
import type * as ProcessFormats from "../processFormats";
import SendMessagesThread from "../sendMessages";
import ReceiveMessagesThread from "../receiveMessages";

export function read(
	receiveThread: ReceiveMessagesThread,
	args: defined[],
	formatData: ProcessFormats.FormatData,
	masks: Map<string, boolean>,
) {
	const [readBuffer, readOffset] = ReceiveMessagesThread.readBuffer(receiveThread, 4);
	const bufLen = buffer.readu32(readBuffer, readOffset);
	const [readBuffer2, readOffset2] = ReceiveMessagesThread.readBuffer(receiveThread, bufLen);
	const bufferValue = buffer.create(bufLen);
	buffer.copy(bufferValue, 0, readBuffer2, readOffset2, bufLen);
	args.push(bufferValue);
}
export function write(
	sendThread: SendMessagesThread,
	args: defined[],
	argsCount: number,
	formatData: ProcessFormats.FormatData,
	masks: Map<string, boolean>,
): number {
	const value = args[argsCount] as buffer;
	const bufLen = buffer.len(value);

	const data = buffer.create(4 + bufLen);
	buffer.writeu32(data, 0, bufLen);
	buffer.copy(data, 4, value);
	SendMessagesThread.writeBuffer(sendThread, data, 4 + bufLen);
	return argsCount + 1;
}
